rust-src/key_derivation/src/lib.rs

```diff
@@ -0,0 +1,256 @@
+use ed25519_hd_key_derivation::derive;
+use keygen_bls::keygen_bls;
+use pairing::bls12_381::FrRepr;
+use serde::Deserialize;
+use std::fmt;
+
+#[derive(Deserialize)]
+pub enum Net {
+    Mainnet,
+    Testnet,
+}
+
+impl fmt::Display for Net {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Net::Mainnet => write!(f, "919"),
+            Net::Testnet => write!(f, "1"),
+        }
+    }
+}
+
+fn bls_key_bytes_from_seed(key_seed: [u8; 32]) -> [u8; 32] {
+    let bls_key_fr = keygen_bls(&key_seed, b"").unwrap();
+    let fr_repr = FrRepr::from(bls_key_fr);
+    let fr_repr_ref = fr_repr.as_ref();
+    let key_bytes: Vec<u8> = fr_repr_ref
+        .iter()
+        .rev()
+        .flat_map(|val| val.to_be_bytes())
+        .collect();
+    let mut bls_key = [0u8; 32];
+    bls_key.clone_from_slice(&key_bytes[0..32]);
+    bls_key
+}
+
+pub struct ConcordiumHdWallet {
+    pub seed: Vec<u8>,
+    pub net:  Net,
+}
+
+impl ConcordiumHdWallet {
+    fn make_path(&self, path: &str) -> String {
+        let root_path: &str = "m/44'";
+        let derivation_path = format!("{}/{}'/{}", root_path, &self.net, path);
+        derivation_path
+    }
+
+    pub fn get_account_signing_key(&self, identity_index: u32, credential_index: u32) -> [u8; 32] {
+        let path = self.make_path(&format!("{}'/0'/{}'", identity_index, credential_index));
+        derive(&path, &self.seed).unwrap().key
+    }
```

Comment:
Haha, I get what you are saying. I haven't even considered that as most of these libraries take the path as a string as the input. The string format is just the readable version of a path, which also quickly tells you what indices are hardened and not. But for ed25519 only hardened indices are allowed, so we don't benefit that much from it.

We could change it so that the library takes a `Vec<u32>` instead of a string. This would also make it possible to get rid of the regex checking, as we wouldn't need it. Someone using the library who wants to use paths in that form would then write the converter instead. I won't be able to redo this today, but I wouldn't mind it being changed.

Reply:
You are correct that `unwrap` is not safe there. I am changing it all to `Result`.

Reply:
I have added a function that allows you to supply a list of u32 values instead of a string, and am using that now in the library to derive prf, idcredsec, account keys, etc.